### ISSUE TITLE
Training speed is very slow

### ISSUE THREAD

**Issue by TIanCat** [XZheng0427/XFMamba#2]
Dear Author,

I recently read your paper XFMamba, and I found it to be a very interesting and promising work. Thank you for sharing the code!

However, I encountered some issues while trying to reproduce your results. Specifically, when training the small model on the MURA dataset with a batch size of 8, the memory usage reaches about 17GB (on an RTX 3090), and it takes around 105 hours to train a single epoch. The GPU utilization is at 100%, so I don't think the bottleneck is due to data loading.

May I ask how long it took in your case to train one epoch? Would it be possible for you to share your training logs for reference? I would greatly appreciate it.

**Comment by TIanCat** [comment]
> Hi,
> 
> When training XFMamba on an A100 (40GB), one epoch on the MURA dataset takes approximately 14 minutes based on the training logs. Could you please double-check your configuration or share more details about your setup/configuration?
> 
> Thanks

Thank you very much for your reply!
I’m currently reviewing my training configuration and will share more detailed information shortly.
May I ask what batch size you used during training, and approximately how much GPU memory was consumed? The reason I ask is that in my reproduction, the memory usage already reaches around 17GB with just batch size = 8, which doesn’t seem quite reasonable.
I’d really appreciate any further clarification or suggestions!

**Comment by hhhhhhharui** [comment]
Hello!

Here are some details regarding the issue mentioned above. I trained the model on an NVIDIA GeForce RTX 3090 with CUDA version 11.8. I first trained the proposed small model from your paper and it turned out that it would take about 100 hours. When using ResNet-18 as the backbone and retaining the two Mamba modules for fusion, as described in your paper, training on the MURA dataset takes approximately one hour. However, after replacing the fusion module with a simpler method that directly combines the features from the two views, the training time was significantly reduced to around 3 minutes.

The environment includes PyTorch 2.1.2, Python 3.10.18, NumPy 1.24.4, and TIMM 0.4.12. We set the batch size to 8 in both cases.

Could you share any insights or suggestions regarding this performance difference? Thank you in advance for your guidance.

**Comment by TIanCat** [comment]
> Hi,
> 
> Could you please check your 'num_workers' and 'pin_memory'? Also, could you please check your Mamba/SSM env installation correctly? Below is the configuration:
> 
> def get_config(args): """Get configuration dictionary for model training""" return { 'root_dir': Path(args.root_dir), 'train_image_paths': Path(args.train_image_paths), 'train_image_labels': args.train_image_labels, 'valid_image_paths': args.valid_image_paths, 'valid_image_labels': args.valid_image_labels, 'confusion_matrix_folder': args.confusion_matrix_folder, 'dataset': args.dataset, 'view_num': args.view_num, 'view_sel': args.view_sel, 'train_ratio': 0.85, 'valid_ratio': 0.15, 'image_size': (224, 224), 'batch_size': args.batch_size, 'num_workers': 4, 'pin_memory': True, 'seed': 42, 'num_classes': args.num_classes, #14 'num_tasks': args.num_tasks, #4 'learning_rate': args.lr, #1e-4, 'weight_decay': 0.01, 'betas':(0.9,0.999), 'weight_decay': 0.01, 'num_epochs': args.epochs, 'early_stopping_patience': 100, 'device': torch.device("cuda" if torch.cuda.is_available() else "cpu"),
> 
> ```
> }
> ```

Thank you again for your response.

We have been following your settings exactly — we cloned your repository and did not modify any configurations. As my colleague mentioned above, we have identified that the backbone model itself seems to be the main reason for the slow training.

Here are our findings:

(1) Using the small model results in a training time of ~105 hours per epoch.

(2) Switching to the tiny model reduces this to ~64 hours per epoch.

(3) Replacing your backbone with ResNet18, while keeping your shallow and deep fusion modules, further reduces the training time to ~1 hour per epoch.

(4) Finally, if we also replace the fusion modules with a simple average operation between the two feature levels, the training time drops to ~3 minutes per epoch.

From these experiments, it seems clear that the model architecture itself is the main cause of the extremely slow training.

Additionally, as mentioned before, the memory usage reaches ~17GB (small model) and 7GB (tiny model) with batch size = 8, which is unusually high and difficult to work with.

We’re continuing to investigate, but we wanted to share these observations and would appreciate any further insights or suggestions you might have.

Thank you again for your time and support!

**Comment by TIanCat** [comment]
> This is my training log for your reference:
> 
> <img alt="Image" width="550" height="74" src="https://private-user-images.githubusercontent.com/97288936/473046366-b68a86e1-9faf-4c90-82ac-9377a6cbd855.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM5NzUyNTQsIm5iZiI6MTc1Mzk3NDk1NCwicGF0aCI6Ii85NzI4ODkzNi80NzMwNDYzNjYtYjY4YTg2ZTEtOWZhZi00YzkwLTgyYWMtOTM3N2E2Y2JkODU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzMxVDE1MTU1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY4YWRmNjJlZDk3ZDAyZjRmNmUwZWQ2N2Y5NzVhMWMwMzM1Y2FjNzM1ZGU4ZWFhZDIwMDhlNDNhOTBjNWM5MzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UFed01AFhJ0t0z54cwlnhD3o8LRWWMqZqwhADS28Brk">

Thank you for providing additional training details!

I’ll proceed to run VMamba as a next step to help determine whether the extremely slow training might be related to hardware or driver issues on my side.

That said, beyond the training speed, I’m also quite concerned about the high GPU memory usage. If possible, could you share how much GPU memory was used during your training with the small model?

Thank you again for your kind support and assistance!

**Comment by yooinsung** [comment]
> > This is my training log for your reference:
> > <img alt="Image" width="550" height="74" src="https://private-user-images.githubusercontent.com/97288936/473046366-b68a86e1-9faf-4c90-82ac-9377a6cbd855.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQwMjIwMjEsIm5iZiI6MTc1NDAyMTcyMSwicGF0aCI6Ii85NzI4ODkzNi80NzMwNDYzNjYtYjY4YTg2ZTEtOWZhZi00YzkwLTgyYWMtOTM3N2E2Y2JkODU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODAxVDA0MTUyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlMjMwNzc1OTRlYjRmYmEyZWRkOTc5MTMxYTU3YmIwZjcxNGNkZGJjZjBkNDUxYzYxOGRmMzdkMTgxZTBiYTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cXtzfjCUdNf-Dg0lArpjlERt4svWqFJurBKuJaDg2Vk">
> 
> Hi,
> 
> We’ve identified the cause of the slow training and have successfully resolved it — the training speed is now back to normal. Thank you again for your kind help and support!

plz share the solution. I have the same problem too

**Comment by qaixerabbas** [comment]
Dear @TIanCat 
Could you please share your insights on model training speed? I am trying to train and facing the same speed issues.
